vpr/src/place/annealer.cpp

```diff
@@ -0,0 +1,898 @@
+         * their original positions (reverting blk_loc_registry.block_locs to its original state).
+         *
+         * Note that the inverse look-up place_ctx.grid_blocks is only updated after
+         * move acceptance is determined, so it should not be used when evaluating a move.
+         */
+
+        /* Update the block positions */
+        blk_loc_registry.apply_move_blocks(blocks_affected_);
+
+        //Find all the nets affected by this swap and update the wiring costs.
+        //This cost value doesn't depend on the timing info.
+        //
+        //Also find all the pins affected by the swap, and calculates new connection
+        //delays and timing costs and store them in proposed_* data structures.
+        net_cost_handler_.find_affected_nets_and_update_costs(delay_model_, criticalities_, blocks_affected_,
+                                                              bb_delta_c, timing_delta_c);
+
+        //For setup slack analysis, we first do a timing analysis to get the newest
+        //slack values resulted from the proposed block moves. If the move turns out
+        //to be accepted, we keep the updated slack values and commit the block moves.
+        //If rejected, we reject the proposed block moves and revert this timing analysis.
+        if (place_algorithm == e_place_algorithm::SLACK_TIMING_PLACE) {
+            // Invalidates timing of modified connections for incremental timing updates.
+            pin_timing_invalidator_->invalidate_affected_connections(blocks_affected_, timing_info_);
+
+            /* Update the connection_timing_cost and connection_delay
+             * values from the temporary values. */
+            placer_state_.mutable_timing().commit_td_cost(blocks_affected_);
+
+            /* Update timing information. Since we are analyzing setup slacks,
+             * we only update those values and keep the criticalities stale
+             * so as not to interfere with the original timing driven algorithm.
+             *
+             * Note: the timing info must be updated after applying block moves
+             * and committing the timing driven delays and costs.
+             * If we wish to revert this timing update due to move rejection,
+             * we need to revert block moves and restore the timing values. */
+            criticalities_->disable_update();
+            setup_slacks_->enable_update();
+            update_timing_classes(crit_params, timing_info_, criticalities_,
+                                  setup_slacks_, pin_timing_invalidator_, placer_state_);
+
+            /* Get the setup slack analysis cost */
+            //TODO: calculate a weighted average of the slack cost and wiring cost
+            delta_c = analyze_setup_slack_cost(setup_slacks_, placer_state_) * costs_.timing_cost_norm;
+        } else if (place_algorithm == e_place_algorithm::CRITICALITY_TIMING_PLACE) {
```

Comment:
This is the default code path, so moving it to be first in the if would be slightly more clear.

Reply:
Can also say this placement algorithm (CRITICALITY_TIMING_PLACE) mixes timing and wiring terms, and works with somewhat stale timing data (periodically updated) to save CPU time.